web_generator/lib/src/interop_gen/transform/transformer.dart

```diff
@@ -1072,34 +1229,75 @@ class Transformer {
         if (isHomogenous) {
           if (nonNullLiteralTypes.isNotEmpty && onlyContainsBooleanTypes) {
             return BuiltinType.primitiveType(PrimitiveType.boolean,
-                isNullable: isNullable);
+                isNullable: shouldBeNullable);
           }
 
-          final expectedId =
-              ID(type: 'type', name: types.map((t) => t.id.name).join('|'));
+          final idMap = nonNullLiteralTypes.map((t) => t.value.toString());
+
+          final expectedId = ID(type: 'type', name: idMap.join('|'));
 
           if (typeMap.containsKey(expectedId.toString())) {
             return typeMap[expectedId.toString()] as UnionType;
           }
 
-          final (id: _, name: name) =
-              namer.makeUnique('AnonymousUnion', 'type');
+          final name =
+              'AnonymousUnion_${AnonymousHasher.hashUnion(idMap.toList())}';
 
           // TODO: Handle similar types here...
-          final homogenousEnumType = HomogenousEnumType(
-              types: nonNullLiteralTypes, isNullable: isNullable, name: name);
+          final homogenousEnumType =
+              HomogenousEnumType(types: nonNullLiteralTypes, name: name);
+
+          final homogenousEnum = typeMap.putIfAbsent(expectedId.toString(), () {
+            namer.markUsed(name);
+            return homogenousEnumType;
+          }) as HomogenousEnumType;
+
+          return homogenousEnum..isNullable = shouldBeNullable;
+        } else {
```

Comment:
There's a lot of duplication in the `else` branch. Can we move the shared code outside of the branch instead?